### ISSUE TITLE
Using Spyder with Different Environments

### ISSUE THREAD

**Issue by Maynaq** [spyder-ide/spyder#15881]
## Problem Description
_Sorry if this is not the right place to write a suggestion-ish post, I could not find the best place to post it._
**Not a problem exactly, more like a suggestion.**
I am using Spyder with Anaconda and wanted to use it with different environments. I decided to use the **modular approach**, which works fine. However, I have two problems\requests with it. 

1. Each time I switch between environments, activating the environment from command prompt then executing the following command _python -c "import sys; print(sys.executable)"_ is a bit exhaustive (File explore in Python interpreter works fine, if all the environments are in the same directory like ../anaconda3/envs/). Instead it would be nice to have a saved list of the old interpreters used. Having this selection on the console (right below) wouldn't hurt either.
 
2. When I start a new console with a different environment, I would like to know which environment I am working with. It is again a bit exhaustive to check the interpreter every time if I am on the right environment or not. So when giving the info message, an additional information about the environment can be given. 

![image](https://user-images.githubusercontent.com/8599461/122179697-70528200-ce90-11eb-8d58-9104c102974a.png)

**Current info message:** 
Python 3.8.10 (default, May 19 2021, 13:12:57) [MSC v.1916 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython  -- An enhanced Interactive Python.

**An example info message:** 
Python 3.8.10 (default, May 19 2021, 13:12:57) [MSC v.1916 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython  -- An enhanced Interactive Python.
**Current interpreter -- C:\Users\theuser\anaconda3\envs\myEnv\python.exe**

**or**

**the active environment -- myEnv**

### What steps reproduce the problem?

I used the following two links to use my own virtual environment with Spyder.
1. https://stackoverflow.com/questions/30170468/how-to-run-spyder-in-virtual-environment
2. https://github.com/spyder-ide/spyder/wiki/Working-with-packages-and-environments-in-Spyder#working-with-other-environments-and-python-installations

### What is the expected output? What do you see instead?
After Spyder has started, navigate to Preferences > Python Interpreter > Use the following interpreter and paste the path from Step 3 into the text box.

## Versions

* Spyder version: 5.0.3
* Python version: 3.8.5
* Qt version: 5.9.7
* PyQt version: 5.9.2
* Operating System name/version: Windows 10

### Dependencies

```

# Mandatory:
atomicwrites >=1.2.0          :  1.4.0 (OK)
chardet >=2.0.0               :  4.0.0 (OK)
cloudpickle >=0.5.0           :  1.6.0 (OK)
cookiecutter >=1.6.0          :  1.7.2 (OK)
diff_match_patch >=20181111   :  20200713 (OK)
intervaltree >=3.0.2          :  3.1.0 (OK)
IPython >=7.6.0               :  7.22.0 (OK)
jedi =0.17.2                  :  0.17.2 (OK)
jsonschema >=3.2.0            :  3.2.0 (OK)
keyring >=17.0.0              :  23.0.1 (OK)
nbconvert >=4.0               :  6.0.7 (OK)
numpydoc >=0.6.0              :  1.1.0 (OK)
paramiko >=2.4.0              :  2.7.2 (OK)
parso =0.7.0                  :  0.7.0 (OK)
pexpect >=4.4.0               :  4.8.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=5.3                  :  5.8.0 (OK)
pygments >=2.0                :  2.9.0 (OK)
pylint >=1.0                  :  2.8.3 (OK)
pyls >=0.36.2;<1.0.0          :  0.36.2 (OK)
pyls_black >=0.4.6            :  0.4.6 (OK)
pyls_spyder >=0.3.2;<0.4.0    :  0.3.2 (OK)
qdarkstyle =3.0.2             :  3.0.2 (OK)
qstylizer >=0.1.10            :  0.1.10 (OK)
qtawesome >=1.0.2             :  1.0.2 (OK)
qtconsole >=5.1.0             :  5.1.0 (OK)
qtpy >=1.5.0                  :  1.9.0 (OK)
rtree >=0.9.7                 :  0.9.7 (OK)
setuptools >=39.0.0           :  52.0.0.post20210125 (OK)
sphinx >=0.6.6                :  4.0.1 (OK)
spyder_kernels >=2.0.3;<2.1.0 :  2.0.3 (OK)
textdistance >=4.2.0          :  4.2.1 (OK)
three_merge >=0.1.1           :  0.1.1 (OK)
watchdog >=0.10.3;<2.0.0      :  1.0.2 (OK)
zmq >=17                      :  20.0.0 (OK)

# Optional:
cython >=0.21                 :  0.29.23 (OK)
matplotlib >=2.0.0            :  3.3.4 (OK)
numpy >=1.7                   :  1.20.2 (OK)
pandas >=1.1.1                :  1.2.4 (OK)
scipy >=0.17.0                :  1.6.2 (OK)
sympy >=0.7.3                 :  1.8 (OK)
```

**Comment by rhkarls** [comment]
Just to comment on your point 1., it might be that the functionality you are after already exists - it's nevertheless good ideas you bring up. Perhaps I misunderstand how you do this but you don't need to activate the (conda) environment in the command prompt to use it in Spyder. Are you familiar with the built in python interpreter selection? It also shows the active environment in the status bar and saves previous environments (also auto-detects conda environments) - I think this works very well. See https://docs.spyder-ide.org/current/faq.html#using-existing-environment . A caveat is if you change the setting and start a new console, it is not possible to easily see what environment _previously_ existing consoles are in, it only shows the current setting. You can right click on the console tabs and the contents of `show environmental variables` and `show sys.path contents` will tell you the environment of that console - as an alternative to `import sys; sys.executable`. You can probably also create your own IPython magic command for a shortcut, and at least if you use conda there is `%env CONDA_ENV_PYTHON`. 

 I have to admit that I seldom jump back and forth between environments when I work though.

**Comment by Maynaq** [comment]
@rhkarls yeah you are right, it already exits. its my bad for not searching it properly. I just got into the interpreter selection and I thought it did not saved the previous environments. So yeah, the first one is already implemented. The commands that you provided are indeed helpful to learn the environment that you are currently using, which I just learned. 

I can say that I don't switch between environments that much as well, and I think thats why the 2nd suggestion would be nice to have. Since I don't jump around environments, I can switch to one environment and completely forget about being in that one. Whenever I start spyder, the first thing I see is the message in the console. It would be a nice reminder to see the latest environment used since the current interpreter will be from the latest environment activated.

**Comment by rhkarls** [comment]
@Maynaq Glad my comment helped you! You can see the name of the conda environment in the status bar as `conda: envname (python version)` and if you hover on the name you see the path of the executable, if that also helps. There is also settings and `IPython console` -> `Startup`. It should be possible to run lines or a script at the startup of each console, for example `import sys; print(f"Current interpreter: {sys.executable}")`, however the IPython console does not support print statements like that at startup it seems.

**Comment by Maynaq** [comment]
@rhkarls once again you saved the day, it seems the features are already there, I just did not know where to look at. I think these are more than sufficient. Thanks a lot!

**Comment by rhkarls** [comment]
> 
> 
> @rhkarls once again you saved the day, it seems the features are already there, I just did not know where to look at. I think these are more than sufficient. Thanks a lot!

No need to thank me, thank the people developing it :D

**Comment by Maynaq** [comment]
@rhkarls sorry to bug you again. As you said `import sys; print(f"Current interpreter: {sys.executable}")` should work and it sometimes work. 
![image](https://user-images.githubusercontent.com/8599461/122217385-0947c400-ceb6-11eb-9ceb-23ba4f520240.png)
This is just a fresh spyder opened with the startup codes:
![image](https://user-images.githubusercontent.com/8599461/122217488-2381a200-ceb6-11eb-9d73-2cf3f14a564c.png)

Not only the print option is not working, also the import sys is not run as well. I wonder if its how it supposed to be or am I missing something.


When I put the lines again it asks me the following:
![image](https://user-images.githubusercontent.com/8599461/122217789-6e031e80-ceb6-11eb-9065-5d5c92757dfb.png)

if I say keep existing kernels, the new consoles does not print out or import sys. However, if I click one of the apply options, it prints out.

![image](https://user-images.githubusercontent.com/8599461/122218041-ab67ac00-ceb6-11eb-9675-a736208b1864.png)

So I wonder if this is intended, or I am in the wrong, or something is wrong 😄

**Comment by rhkarls** [comment]
My bad for suggestion those particular start-up lines and not making it clear that the IPython console currently does _not_ support print statements like that at startup. I also tested to `import sys` and as you said notice it's not imported - and the reason appears that its later deleted (`del sys` is executed later in the startup internally in `spyder\external-deps\spyder-kernels\spyder_kernels\console\start.py` - a separate bug if that is the case).

However, I found a little workaround for your print issue and this works for me (add this to start up lines as you did above):

    import sys; get_ipython().banner1='\nCurrent old interpreter: ' + sys.executable

You can make it prettier of course, but this prints the path to the current interpreter each time a console is opened. You can also add something to get_ipython().banner2. They both appear blank by default, but if Spyder devs decide to use these banners you might overwrite something. Perhaps someone with more knowledge how this works internally can chime in

**Comment by dalthviz** [comment]
Hi @Maynaq and @rhkarls thank you for the feedback and checking and discussing a work around! Pinging @ccordoba12 just in case since I think some work regarding the interpreter config will be done in the future

**Comment by ccordoba12** [comment]
> I am using Spyder with Anaconda and wanted to use it with different environments

This is a duplicate of issue https://github.com/spyder-ide/ux-improvements/issues/10 and we'll try to implement it in the coming months.

> As you said import sys; print(f"Current interpreter: {sys.executable}") should work and it sometimes work.

The problem with the sys module was fixed in our 5.0.4 version, released two weeks ago. Please update.
